### PR TITLE
Fix resolves issue #6366 updated

### LINKED ISSUE #6366: DescriptorMatcher::checkMasks fails on Nx1 masks
### Please state the information for your system
- OpenCV version: 3.x
- Host OS: Linux (Ubuntu 14.04)

### In which part of the OpenCV library you got the issue?
- feature2d
- matcher

### Expected behaviour
Running knnMatch or radiusMatch with a Nx1 mask pass on the assertion inside DescriptorMatcher::checkMasks method.

### Actual behaviour
Running fails with the following error
> 
matchers.cpp:617: error: (-215) masks[i].rows == queryDescriptorsCount && (masks[i].cols == rows || masks[i].cols == rows) && masks[i].type() == CV_8UC1 in function checkMasks

### Additional description
The assertion fails becaus the second condition of the assertion only compares the number of columns of the masks.

### Code example to reproduce the issue / Steps to reproduce the issue
```c++
#include <iostream>

#include <string>
#include <vector>

#include <opencv2/highgui.hpp>
#include <opencv2/videoio.hpp>

#include <opencv2/features2d.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/video/video.hpp>

int main(int argc, char** argv) {

    std::string folder("");

    std::vector<cv::Mat> images;
    images.push_back(cv::imread("/home/anishitani/workspace/visionws/src/tutorial/data/fountain/0000.jpg",cv::IMREAD_GRAYSCALE));
    images.push_back(cv::imread("/home/anishitani/workspace/visionws/src/tutorial/data/fountain/0001.jpg",cv::IMREAD_GRAYSCALE));

    /** Feature detection init **/
    std::vector<std::vector<cv::KeyPoint> > fkpts(2);

    cv::Ptr<cv::FeatureDetector> detector = cv::FastFeatureDetector::create();
    detector->detect(images[0],fkpts[0]);
    detector->detect(images[1],fkpts[1]);
    /** Feature detection end **/

    /** Feature matching init **/
    std::vector<std::vector<cv::Point2f> > fpts(2);
    cv::KeyPoint::convert(fkpts[0],fpts[0]);
    cv::KeyPoint::convert(fkpts[1],fpts[1]);
    // Calculate the optical flow field:
    // how each left_point moved across the 2 images
    std::vector<uchar> vstatus;
    std::vector<float> verror;
    cv::calcOpticalFlowPyrLK(images[0], images[1], fpts[0], fpts[1], vstatus, verror);
    /** Feature matching end **/

    cv::Mat out;
    cv::cvtColor(images[0],out,cv::COLOR_GRAY2BGR);
    std::vector<cv::Point2f> flowpts;
    for(int i=0 ; i<fpts[0].size() ; i++){
        if(vstatus[i] && verror[i]<12.0){
            flowpts.push_back(fpts[1][i]);
        } else {
            flowpts.push_back(cv::Point2f());
            vstatus[i] = 0;
        }
    }

    cv::BFMatcher matcher;
    std::vector<std::vector<cv::DMatch> > nn;
    cv::Mat mask;
    cv::Mat(vstatus).convertTo(mask,CV_8UC1);

    matcher.radiusMatch( cv::Mat(fpts[1]).reshape(1,fpts[1].size()),
            cv::Mat(flowpts).reshape(1,flowpts.size()),
            nn,2.0f, mask);

    cv::namedWindow("features",cv::WINDOW_KEEPRATIO);
    cv::imshow("features",out);

    for(char k = cv::waitKey() ; k!='q' ; k = cv::waitKey());

    return 0;
}

```

### PR DESCRIPTION
**Pull request by souch55**
Resolves #6366

Replaces #7099

### CODE REVIEW

File: modules/features2d/src/matchers.cpp [hunk 1]
```diff
@@ -601,22 +601,32 @@ void DescriptorMatcher::checkMasks( InputArrayOfArrays _masks, int queryDescript
...
         for( size_t i = 0; i < imageCount; i++ )
-        {
-            if( !masks[i].empty() && (!trainDescCollection[i].empty() || !utrainDescCollection[i].empty() ) )
-            {
-                int rows = trainDescCollection[i].empty() ? utrainDescCollection[i].rows : trainDescCollection[i].rows;
+        { 
+            if(!masks[i].empty()){
+        	  if(!trainDescCollection[i].empty() || !utrainDescCollection[i].empty())
+        	  { 
+                    rows = trainDescCollection[i].rows;
```
Comment (terfendail) [inline modules/features2d/src/matchers.cpp]: Execution will fall into this branch if either trainDescCollection or utrainDescCollection isn't empty. So rows should be populated in accordance.

File: modules/features2d/src/matchers.cpp [hunk 2]
```diff
@@ -601,22 +601,32 @@ void DescriptorMatcher::checkMasks( InputArrayOfArrays _masks, int queryDescript
...
-                        (masks[i].cols == rows || masks[i].cols == rows) &&
-                        masks[i].type() == CV_8UC1 );
-            }
-        }
+        	                          (masks[i].cols == rows || masks[i].rows == rows) &&
+        	                           masks[i].type() == CV_8UC1 );
+                   }
+        	  else
+        	  { 
+                    rows = utrainDescCollection[i].rows;
```
Comment (terfendail) [inline modules/features2d/src/matchers.cpp]: Looks like this check should be a part of other branch(or other condition should be used)

**Comment by terfendail** [comment]
Could you please also fix whitespacing [issues](http://pullrequest.opencv.org/buildbot/builders/precommit_docs/builds/6773/steps/whitespace%20opencv/logs/stdio)?

**Comment by catree** [comment]
I don't know if it fixes the issue but I think your code:

```
 if(!masks[i].empty()){
        	  if(!trainDescCollection[i].empty() || !utrainDescCollection[i].empty())
        	  { 
                    rows = trainDescCollection[i].rows;
                    CV_Assert( masks[i].rows == queryDescriptorsCount &&
        	                          (masks[i].cols == rows || masks[i].rows == rows) &&
        	                           masks[i].type() == CV_8UC1 );
                   }
        	  else
        	  { 
                    rows = utrainDescCollection[i].rows;
        	    CV_Assert( masks[i].rows == queryDescriptorsCount &&
        	                          (masks[i].cols == rows || masks[i].rows == rows) &&
        	                          masks[i].type() == CV_8UC1 ); 
        	 }
              } 
```

should be:
```
            if( !masks[i].empty() && (!trainDescCollection[i].empty() || !utrainDescCollection[i].empty() ) )
            {
                int rows = trainDescCollection[i].empty() ? utrainDescCollection[i].rows : trainDescCollection[i].rows;
                    CV_Assert( masks[i].rows == queryDescriptorsCount &&
                        (masks[i].cols == rows || masks[i].rows == rows) &&
                        masks[i].type() == CV_8UC1 );
            }
```
as in the original [code](https://github.com/opencv/opencv/blob/3.1.0/modules/features2d/src/matchers.cpp#L612) but with `masks[i].rows == rows` instead of twice the comparison `masks[i].cols == rows`.

**Comment by souch55** [comment]
Hi catee both the fixes are correct as tested ,yes there were two times comparison of same things masks[i].cols == rows,i have created this new pull request please have an look #https://github.com/opencv/opencv/pull/7114

**Comment by mshabunin** [comment]
Do we really need to compare `masks[i].rows == rows`? As @ohnozzy [wrote](https://github.com/opencv/opencv/issues/6366#issuecomment-207354286) in original issue (#6366), and as [documentation ](http://docs.opencv.org/master/db/d39/classcv_1_1DescriptorMatcher.html#a56b8dcd6fc600f8f2a23212f6e4c5082) states: 

>Mask specifying permissible matches between an input query and train matrices of descriptors

So each mask should have number of rows equal to number of query descriptors and number of columns equal to number of train descriptors.

**Comment by mshabunin** [comment]
The PR has been recreated.

**Pull request closed without merging**